File: 0402-remove-k-digits/0402-remove-k-digits.py
from collections import deque
class Solution(object):
    def removeKdigits(self, num, k):
        """
        :type num: str
        :type k: int
        :rtype: str
        """
        stack = []
        for i in num:
            while(stack and k>0 and stack[-1]>i):
                k -= 1
                stack.pop()
            stack.append(i)
        stack = stack[:-k] if(k>0) else stack
        s = ''.join(stack)
        s = s.lstrip('0')
        return "0" if(len(s)==0) else s
        

File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        stack<char> st;
        for(char i : num){
            while(!st.empty() && k > 0 && i < st.top()){
                st.pop();
                k--;
            }
            st.push(i);
        }
        while(!st.empty() && k){
            st.pop();
            k--;
        }
        string res;
        while(!st.empty()){
            res += st.top();
            st.pop();
        }
        int i = 0;
        reverse(res.begin(),res.end());
        while(res[i] == '0') i++;
        res = res.substr(i);
        return res.size() == 0 ? "0" : res;
    }
};